### PR TITLE
Libnetwork revendoring

### LINKED ISSUE #32093: ndots:0
Docker imposes ndots:0 in current resolv.conf
This conflicts with search option and prevent containers from finding nodes in an intranet using short names.
Removing the ndots:0 option fixes that, so there should be a way to override it. Providing a different ndots value currently results in:
```
cat /etc/resolv.conf
search mynetwork.net
nameserver 127.0.0.11
options ndots:15 ndots:0
```

### PR DESCRIPTION
**Pull request by fcrisciani**
**- What I did**
Diff:
https://github.com/docker/libnetwork/compare/5ab4ab830062fe8a30a44b75b0bda6b1f4f166a4...20dd462e0a0e883437a274bd61df4bc4de980830

**- Description for the changelog**

- https://github.com/docker/libnetwork/pull/2040 Memberlist revendor (fix for deadlock on exit)
- https://github.com/docker/libnetwork/pull/2032 Network diagnostic client
- https://github.com/docker/libnetwork/pull/2065 Fix for ndots configuration

Fixes https://github.com/moby/moby/issues/32093

### CODE REVIEW

**Review by vdemeester — APPROVED** [review]
LGTM 🐯

File: daemon/reload.go [hunk 1]
```diff
@@ -312,17 +312,22 @@ func (daemon *Daemon) reloadLiveRestore(conf *config.Config, attributes map[stri
...
+	}
+	if conf == nil {
+		// If there is no config make sure that the diagnostic is off
+		daemon.netController.StopDiagnostic()
 		return nil
 	}
-	// Enable the network diagnose if the flag is set with a valid port withing the range
+	// Enable the network diagnostic if the flag is set with a valid port withing the range
 	if conf.IsValueSet("network-diagnostic-port") && conf.NetworkDiagnosticPort > 0 && conf.NetworkDiagnosticPort < 65536 {
 		logrus.Warnf("Calling the diagnostic start with %d", conf.NetworkDiagnosticPort)
```
Comment (cpuguy83) [inline daemon/reload.go]: Why is this a warning?

File: daemon/reload.go [hunk 2]
```diff
@@ -312,17 +312,22 @@ func (daemon *Daemon) reloadLiveRestore(conf *config.Config, attributes map[stri
...
-		daemon.netController.StartDiagnose(conf.NetworkDiagnosticPort)
+		daemon.netController.StartDiagnostic(conf.NetworkDiagnosticPort)
 	} else {
-		daemon.netController.StopDiagnose()
+		daemon.netController.StopDiagnostic()
```
Comment (cpuguy83) [inline daemon/reload.go]: Can we move this up and return early instead of using else?

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

File: daemon/reload.go (hunk 1 above)
Comment (fcrisciani) [reply daemon/reload.go]: Because is a special diagnostic function that has security implication we want to give a clear notice of it to the user

Comment (fcrisciani) [reply daemon/reload.go]: Is this looking better?
```
func (daemon *Daemon) reloadNetworkDiagnosticPort(conf *config.Config, attributes map[string]string) error {
	if conf == nil || daemon.netController == nil || !conf.IsValueSet("network-diagnostic-port") || conf.NetworkDiagnosticPort < 1 || conf.NetworkDiagnosticPort > 65535 {
		// If there is no config make sure that the diagnostic is off
		if daemon.netController != nil {
			daemon.netController.StopDiagnostic()
		}
		return nil
	}
	// Enable the network diagnostic if the flag is set with a valid port withing the range
	logrus.Warnf("Calling the diagnostic start with %d", conf.NetworkDiagnosticPort)
	daemon.netController.StartDiagnostic(conf.NetworkDiagnosticPort)

	return nil
}
```

File: daemon/reload.go (hunk 2 above)
Comment (fcrisciani) [reply daemon/reload.go]: Is this looking better?
```
func (daemon *Daemon) reloadNetworkDiagnosticPort(conf *config.Config, attributes map[string]string) error {
	if conf == nil || daemon.netController == nil || !conf.IsValueSet("network-diagnostic-port") || conf.NetworkDiagnosticPort < 1 || conf.NetworkDiagnosticPort > 65535 {
		// If there is no config make sure that the diagnostic is off
		if daemon.netController != nil {
			daemon.netController.StopDiagnostic()
		}
		return nil
	}
	// Enable the network diagnostic if the flag is set with a valid port withing the range
	logrus.Warnf("Calling the diagnostic start with %d", conf.NetworkDiagnosticPort)
	daemon.netController.StartDiagnostic(conf.NetworkDiagnosticPort)

	return nil
}
```

Comment (cpuguy83) [reply daemon/reload.go]: Sure does. Is there some other validation on these port values?

File: daemon/reload.go (hunk 1 above)
Comment (cpuguy83) [reply daemon/reload.go]: Ok. Maybe we should reword this.
`Starting network diagnostic service on <addr>`

File: daemon/reload.go (hunk 2 above)
Comment (fcrisciani) [reply daemon/reload.go]: like?

File: daemon/reload.go (hunk 1 above)
Comment (fcrisciani) [reply daemon/reload.go]: the address is hardcoded to 127.0.0.1

File: daemon/reload.go (hunk 2 above)
Comment (cpuguy83) [reply daemon/reload.go]: Like sending an error when the ports are invalid?

Comment (fcrisciani) [reply daemon/reload.go]: I checked there is no specific validation done on the interval of the value itself, only validates that is a number, as for now. We kept it simple being a config that is only for advanced users.

**Comment by fcrisciani** [comment]
@cpuguy83 is it ok now the patch?

@nishanttotla @anshulpundir  @dperny failures are on the swarmkit side:
Janky:
```
22:52:44 FAIL: docker_cli_swarm_test.go:1855: DockerSwarmSuite.TestSwarmServiceLsFilterMode
22:52:44 
22:52:44 [d8b8d3e6e18c9] waiting for daemon to start
22:52:44 [d8b8d3e6e18c9] daemon started
22:52:44 
22:52:44 docker_cli_swarm_test.go:1876:
22:52:44     c.Assert(out, checker.Not(checker.Contains), "top1")
22:52:44 ... obtained string = "" +
22:52:44 ...     "ID                  NAME                MODE                REPLICAS            IMAGE               PORTS\n" +
22:52:44 ...     "q1top1rns0wk        top2                global              1/1                 busybox:latest      \n"
22:52:44 ... substring string = "top1"
22:52:44 
22:52:44 [d8b8d3e6e18c9] exiting daemon
22:52:51 
22:52:51 ----------------------------------------------------------------------
```

z:
```
22:36:53 ----------------------------------------------------------------------
22:36:53 FAIL: docker_api_swarm_service_test.go:95: DockerSwarmSuite.TestAPISwarmServicesMultipleAgents
22:36:53 
22:36:53 [d0cd66cd10596] waiting for daemon to start
22:36:53 [d0cd66cd10596] daemon started
22:36:53 
22:36:53 [d7ba78554c10a] waiting for daemon to start
22:36:53 [d7ba78554c10a] daemon started
22:36:53 
22:36:53 [dd21353e367a9] waiting for daemon to start
22:36:53 [dd21353e367a9] daemon started
22:36:53 
22:36:53 [d7ba78554c10a] exiting daemon
22:36:53 docker_api_swarm_service_test.go:119:
22:36:53     waitAndAssert(c, defaultReconciliationTimeout, reducedCheck(sumAsIntegers, d1.CheckActiveContainerCount, d3.CheckActiveContainerCount), checker.Equals, instances)
22:36:53 docker_utils_test.go:465:
22:36:53     c.Assert(v, checker, args...)
22:36:53 ... obtained int = 6
22:36:53 ... expected int = 5
22:36:53 ... output: "014a796be661\ndf8c279f336e\n", output: "c798aedd47f4\n951dc1e9310e\n56d9acd993d3\n188823a9ad21\n"
22:36:53 
22:36:53 [d0cd66cd10596] exiting daemon
22:36:53 [dd21353e367a9] exiting daemon
22:36:53 
22:36:53 ----------------------------------------------------------------------
```

**Comment by nishanttotla** [comment]
The second test failure looks like one more instance than expected running.

@fcrisciani do you know what versions these test machines are running?

cc @abhi perhaps this is an instance of the failure we've been tracking down.

**Comment by fcrisciani** [comment]
ping @cpuguy83 @vdemeester @thaJeztah

**Comment by thaJeztah** [comment]
Not sure what version is running on the s390 nodes

**Comment by fcrisciani** [comment]
@nishanttotla tests covers the daemon coming from this build, are you referring to the OS version?

**Comment by thaJeztah** [comment]
Looks like there was still a request to change the message that's printed; https://github.com/moby/moby/pull/36137#discussion_r164564282

**Comment by nishanttotla** [comment]
@fcrisciani I meant daemon version. But it seems like tests are passing now?

**Comment by fcrisciani** [comment]
@nishanttotla the daemon version is top of moby/moby because it tests this PR.
Was pointing the signature as an heads up, the signature is similar to the missmatch between the desired state and running state

File: daemon/reload.go [hunk 3]
```diff
@@ -312,17 +312,19 @@ func (daemon *Daemon) reloadLiveRestore(conf *config.Config, attributes map[stri
...
 	}
-	// Enable the network diagnose if the flag is set with a valid port withing the range
-	if conf.IsValueSet("network-diagnostic-port") && conf.NetworkDiagnosticPort > 0 && conf.NetworkDiagnosticPort < 65536 {
-		logrus.Warnf("Calling the diagnostic start with %d", conf.NetworkDiagnosticPort)
-		daemon.netController.StartDiagnose(conf.NetworkDiagnosticPort)
-	} else {
-		daemon.netController.StopDiagnose()
-	}
+	// Enable the network diagnostic if the flag is set with a valid port withing the range
+	logrus.Warnf("Calling the diagnostic start with %d", conf.NetworkDiagnosticPort)
```
Comment (cpuguy83) [inline daemon/reload.go]: Generally would be nice to do `logrus.WithField("port", conf.NetworkDiagnosticPort).Warn("Starting network diagnostic server")`.

**Review by cpuguy83 — APPROVED** [review]
LGTM

File: daemon/reload.go (hunk 3 above)
Comment (fcrisciani) [reply daemon/reload.go]: I will rebase and put: `logrus.WithFields(&logrus.Fields{"port": conf.NetworkDiagnosticPort, "ip": "127.0.0.1"}).Warn("Starting network diagnostic server")`

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!

**Pull request merged**